doc/designs/quic-design/quic-fault-injector.md

```diff
@@ -0,0 +1,507 @@
+QUIC Fault Injector
+===================
+
+The OpenSSL QUIC implementation receives QUIC packets from the network layer and
+processes them accordingly. It will need to behave appropriately in the event of
+a misbehaving peer, i.e. one which is sending protocol elements (e.g. datagrams,
+packets, frames, etc) that are not in accordance with the specifications or
+OpenSSL's expectations.
+
+The QUIC Fault Injector is a component within the OpenSSL test framework that
+can be used to simulate misbehaving peers and confirm that that OpenSSL QUIC
+implementation behaves in the expected manner as a result of such misbehaviour.
+
+Typically an individual test will inject one particular misbehaviour (i.e. a
+fault) into an otherwise normal QUIC connection. Therefore the fault injector
+will have to be capable of creating fully normal QUIC protocol elements, but
+also offer the flexibility for a test to modify those normal protocol elements
+as required for the specific test circumstances. The OpenSSL QUIC implementaiton
+in libssl does not offer the capability to send faults since it is designed to
+be RFC compliant.
+
+The QUIC Fault Injector will be external to libssl (it will be in the test
+framework) but it will reuse the standards compliant QUIC implementation in
+libssl and will make use of 3 integration points to inject faults. 2 of these
+integration points will use new callbacks added to libssl. The final integration
+point does not require any changes to libssl to work.
+
+QUIC Integration Points
+-----------------------
+
+### TLS Handshake
+
+Fault Injector based tests may need to inject faults directly into the TLS
+handshake data (i.e. the contents of CRYPTO frames). However such faults may
+need to be done in handshake messages that would normally be encrypted.
+Additionally the contents of handshake messages are hashed and each peer
+confirms that the other peer has the same calculated hash value as part of the
+"Finished" message exchange - so any modifications would be rejected and the
+handshake would fail.
+
+An example test might be to confirm that an OpenSSL QUIC client behaves
+correctly in the case that the server provides incorrectly formatted transport
+parameters. These transport parameters are sent from the server in the
+EncryptedExtensions message. That message is encrypted and so cannot be
+modified by a "man-in-the-middle".
+
+To support this integration point two new callbacks will be introduced to libssl
+that enables modification of handshake data prior to it being encrypted and
+hashed. These callbacks will be internal only (i.e. not part of the public API)
+and so only usable by the Fault Injector.
+
+The new libssl callbacks will be as follows:
+
+```` C
+typedef int (*ossl_statem_mutate_handshake_cb)(unsigned char *msgin,
```

Comment:
Where is the `const`?

Or it is there so as to not violate the constness in the suggested `*msgout = msgin` below?

Reply:
Now added for real. I removed the bit about `*msgout = *msgin` since my fault injector implementation never actually used that capability anyway (no mutation is done simply by copying the data into a new buffer).